File: src/main/java/kroppeb/server/command/Arguments/Resource.java
package kroppeb.server.command.Arguments;

import kroppeb.server.command.StringReader;
import kroppeb.server.command.commands.Summon;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Resource {
	final public String namespace;
	final public String[] path;
	
	public Resource(String namespace, String[] path) {
		this.namespace = namespace;
		this.path = path;
	}
	
	public static Resource read(StringReader reader){
		String ns = reader.readIdentifier();
		List<String> path = new ArrayList<String>();
		if(reader.tryRead(':')){
			do{
				path.add(reader.readIdentifier());
			}while(reader.tryRead(':'));
		}else{
			path.add(ns);
			ns = null;
			while(reader.tryRead(':'))
				path.add(reader.readIdentifier());
		}
		return new Resource(ns, path.toArray(EmptyStringArray));
	}
	
	@Override
	public String toString() {
		return (namespace==null?"minecraft":namespace) + ':' + String.join("/", path);
	}
	
	final static String[] EmptyStringArray = new String[0];
}

File: src/main/kotlin/kroppeb/server/command/arguments/Resource.kt
/*
 * Copyright (c) 2020 Kroppeb
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
package kroppeb.server.command.arguments

import kroppeb.server.command.reader.ReadFactory
import kroppeb.server.command.reader.Reader
import kroppeb.server.command.reader.ReaderException

class Resource(val namespace: String?, val path: Array<String?>) {
	override fun toString(): String {
		return (namespace ?: "minecraft") + ':' + java.lang.String.join("/", *path)
	}

	companion object:ReadFactory<Resource> {
		override fun Reader.parse(): Resource {
			val ns = readIdentifier()
			return Resource(ns.namespace, ns.path.split("/").toTypedArray())
		}
	}

}
